### PR TITLE
Ml crypto etch fixed point

### PR DESCRIPTION
**Pull request by baykaner**
Extending #1321 
- sets fixed_point_64 as the default DataType in the vm_modules math/ml libraries
- implements normalisation for crypto prediction example
- implements sgd optimiser for etch binding
- various tidying/refactoring

### CODE REVIEW

File: libs/ml/include/ml/dataloaders/crypto_prediction_loaders/crypto_dataloader.hpp [hunk 1]
```diff
@@ -17,93 +17,110 @@
...
-void TensorDataLoader<LabelType, InputType>::Reset()
-{
-  data_cursor_  = 0;
-  label_cursor_ = 0;
-}
-
-}  // namespace dataloaders
-}  // namespace ml
-}  // namespace fetch
+//#pragma once
```
Comment (wkaluza) [inline libs/ml/include/ml/dataloaders/crypto_prediction_loaders/crypto_dataloader.hpp]: Please don't.

**Review by wkaluza — CHANGES_REQUESTED** [review]
Looks ok, but there's a commented out file that should not go in as is; also, the scaler and optimiser modules deserve to have some tests to keep them safe.

File: scripts/etch-language-tests/ml/dataloader.etch [hunk 2]
```diff
@@ -0,0 +1,3 @@
+function main()
+    var data_loader = DataLoader();
+endfunction
```
Comment (wkaluza) [inline scripts/etch-language-tests/ml/dataloader.etch:3]: I don't think this test is pulling its weight, as it's only exercising the constructor, especially given that the other tests in this PR construct an instance of a DataLoader. If there's nothing interesting one can do with a dataloader in isolation, I would get rid of this one.

File: scripts/etch-language-tests/ml/graph.etch [hunk 3]
```diff
@@ -0,0 +1,24 @@
+function main()
```
Comment (wkaluza) [inline scripts/etch-language-tests/ml/graph.etch:1]: This test is missing assertions that behaviour is correct. Etch now contains an `assert` function you can use, with an optional message.

File: scripts/etch-language-tests/ml/optimiser.etch [hunk 4]
```diff
@@ -0,0 +1,41 @@
+function main()
```
Comment (wkaluza) [inline scripts/etch-language-tests/ml/optimiser.etch:1]: This checks the code compiles and does not crash. It would be nice to also know if the optimisers do the right thing, if for no other reason than to protect the behaviour from being accidentally changed in the future. Add assertions, please.

Comment (wkaluza) [inline scripts/etch-language-tests/ml/optimiser.etch]: This only checks that code involving optimisers compiles and does not crash. It would be nice to know that they also do the right thing. Add assertions, please.

File: scripts/etch-language-tests/ml/scaler.etch [hunk 5]
```diff
@@ -0,0 +1,20 @@
+function main()
+
```
Comment (wkaluza) [inline scripts/etch-language-tests/ml/scaler.etch:2]: As above.

**Review by wkaluza — CHANGES_REQUESTED** [review]

**Review by wkaluza — DISMISSED** [review]

**Review by wkaluza — APPROVED** [review]
